mage-ai/mage-ai
# [hw] Add the support for ClickHouse in SQL blocks

# Summary


Integrate with ClickHouse in SQL blocks.

# Tests


This change has been tested with a locally installed ClickHouse on Mac OS. The setup procedure of ClickHouse 
was following [Quick Install](https://clickhouse.com/docs/en/install), i.e., 
```
$ curl https://clickhouse.com/ | sh
$ ./clickhouse server    # in one terminal
$ ./clickhouse client    # in another terminal
```

Verifying the ClickHouse setup with the following SQL command in the Client prompt:
```
:) SELECT 1

SELECT 1

┌─1─┐
│ 1 │
└───┘

1 rows in set. Elapsed: 0.003 sec.

:)
```

The Python library used for connecting with ClickHouse is [clickhouse-connect](https://github.com/ClickHouse/clickhouse-connect) provided by `ClickHouse.com`. More details about the usage of this Python library is available at [Python Integration with ClickHouse Connect](https://clickhouse.com/docs/en/integrations/python) .

Validation through Mage UI was done through the following steps:

1. Initiate and start the local Mage development environment with the following commands:
```
$ ./scripts/init.sh dev_test
$ ./scripts/dev.sh dev_test
```

2. Open the Mage UI at: http://localhost:3000/

3. Create a new Standard (batch) pipeline, and add a SQL `Data loader` block:
<img width="137" alt="Screenshot 2023-04-13 at 10 33 17 PM" src="https://user-images.githubusercontent.com/42320565/231950099-8477735e-2097-42f6-af4e-3e70c19b18bf.png">

4. Add the following settings to `io_config.yaml`:
```
  CLICKHOUSE_DATABASE: default
  CLICKHOUSE_HOST: host.docker.internal
  CLICKHOUSE_INTERFACE: http
  CLICKHOUSE_PASSWORD: null
  CLICKHOUSE_PORT: 8123
  CLICKHOUSE_USERNAME: null
```

5. In the SQL Block editor, choose `Clickhouse` under `Connection` and `Default` under `Profile`, then run the following statement:
```
select 1;
```
Results similar to the following would be returned:
(1) If `Use raw SQL` is checked:
<img width="962" alt="Screenshot 2023-04-13 at 10 26 36 PM" src="https://user-images.githubusercontent.com/42320565/231950954-0a7563b3-b644-486a-9496-a1b37bfff8c6.png">

(2) If `Use raw SQL` is unchecked:
<img width="960" alt="Screenshot 2023-04-13 at 10 27 16 PM" src="https://user-images.githubusercontent.com/42320565/231951098-c1e8b274-821e-42ee-b951-8f38f621df38.png">

6. A more complicated validation with multiple queries could be run as well:
```
CREATE OR REPLACE TABLE test
(
    id UInt64,
    updated_at DateTime DEFAULT now(),
    updated_at_date Date DEFAULT toDate(updated_at)
)
ENGINE = MergeTree
ORDER BY id;

INSERT INTO test (id) Values (1);

SELECT * FROM test;
```
The results similar to the following would be returned:
<img width="1106" alt="Screenshot 2023-04-13 at 10 44 11 PM" src="https://user-images.githubusercontent.com/42320565/231951984-3d1ab36a-6303-48f0-8da6-20f11d809634.png">

cc:

@dy46 @wangxiaoyou1993 @tommydangerous
